docs/markdown/Helm/helm-deployments.md

```diff
@@ -0,0 +1,197 @@
+image: example.com/registry/my-app:latest
+```
+```yaml src/chart/templates/pod.yaml
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my_pod
+  labels:
+    chart: "{{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}"
+spec:
+  containers:
+    - name: my_app
+      # Uses the `image` value entry from the deployment inputs
+      image: {{ .Values.image }}
+```
+```text src/deployment/BUILD
+# Overrides the `image` value for the chart using the target address for the first-party docker image.
+helm_deployment(dependencies=["src/chart"], values={"image": "src/docker"})
+```
+
+> 📘 Docker image references VS Pants' target addresses
+> 
+> You should use typical Docker registry addresses in your Helm charts. Because Helm charts are distributable artifacts and may be used with other tools other than Pants, you should create your charts such that when that chart is being used, all Docker image addresses are valid references to in accessible Docker registries. We recommend that those resources that will be deploying first-party Docker images to follow what is shown in the previous example, allowing to override the value of the image address in the chart.
+> Besides, your deployments can still use off-the-shelf images published with other means, and in those cases you will also be referencing the Docker image address. Usage of Pants' target addresses is intended for your own first-party images because the image reference of those is not known at the time we create the sources (they are computed later).
+
+With this setup we should be able to run `./pants dependencies src/deployment` and Pants should give the following output:
+
+```text
+src/chart
+src/docker
+```
+
+This should work any kind of Kubernetes resource that will result in Docker image being deployed into Kubernetes, like `Deployment`, `StatefulSet`, `ReplicaSet`, `CronJob`, etc. Please get in touch with us in case you find Pants was not capable to infer dependencies in any of your `helm_deployment` targets by either [opening a GitHub issue](https://github.com/pantsbuild/pants/issues/new/choose) or [joining our Slack](doc:getting-help).
+
+> 📘 How the Docker image reference is calculated during deployment?
+> 
+> Pants' will rely on the behaviour of the `docker_image` target when it comes down to generate the final image reference. Since a given image may have more than one valid image reference, **Pants will try to use the first one that is not tagged as `latest`**, falling back into `latest` if none could be found.
+> It's good practice to publish your Docker images using tags other than `latest` and Pants preferred behaviour is to choose those are this guarantees that the _version_ of the Docker image being deployed is the expected one.
+
+Value override files
+--------------------
+
+As seen in the initial examples of the usage of the `helm_deployment` target, it can be seen that some of the source value files used the word `override` in the file name.
```

Comment:
Does this treat files with "override" in the name differently than the rest of the sources?

The StackStorm chart has a builds deployments for a lot of services so it's got a really long values file. So, I break my values files into a bunch of cross cuting concerns like resources.yaml, dns.yaml, annotations.yaml, and image.yaml. in some cases, the order they are passed to helm matters. So, how can I specify what order the values files are passed to helm?

Reply:
well, the usage of the `override` suffix may well be something that comes from my habits, met some other people that did the same, but defo up to debate.

I thought about leaving this out and just use the same order that is given in the `sources` field of the target however this can be problematic when using globs in that field if different computers compute a list of files in a different order.

The logic implemented would sort the value files making those that have the `override` word in their filename be last in the list passed to Helm. In a complex scenario like yours this is unlikely to be helpful if your files have a strict ordering.

Happy to remove this sorting feature and just use the ordering as listed in the target if we are sure that globing will always return files in the same order.

Reply:
What if we did this (combing the override sort with the sources sort):
- preserve the order of entries in `sources`
- sort the file list within each glob, accounting for `override`:
    - alphanumerically sort for files without `override`:
    - alphanumerically sort for files with `override`
- explicitly listed files get sorted based on where they are in sources, even if they are included in the other globs

So in this example:
```suggestion
helm_deployment(name="dev", dependencies=["//src/chart"], sources=["common.yaml", "*.yaml", "dev/*.yaml", "dev/last.yaml"])
```
I would expect values files to be passed to `helm` in this order:
- common.yaml
- a.yaml
- z.yaml
- a-override.yaml
- z-override.yaml
- dev/a.yaml
- dev/z.yaml
- dev/a-override.yaml
- dev/z-override.yaml
- dev/last.yaml

Reply:
We might be able to provide the overrides sorry with just globs too:
- preserve the order of entries in `sources`
- sort the file list within each glob:
    - alphanumerically sort for files
- explicitly listed files get sorted based on where they are in sources, even if they are included in the other globs

So in this example would achieve the same sort with this algorithm+sources as the previous algorithm+sources:
```python
helm_deployment(name="dev", dependencies=["//src/chart"], sources=["common.yaml", "*.yaml", "*override.yaml", "dev/*.yaml",  "dev/*-override.yaml", "dev/last.yaml"])
```

Reply:
got it, that would remove any magic meaning to the `override` suffix, which is not a bad thing.

It's a less trivial implementation though because it needs some sort of set hierarchy for each of the glob patterns, so a glob like `*.yaml` would ignore the files already included by the `*-override.yaml` pattern.

It may be possible to achieve using an instance of `PathGlobs` per pattern such that the other patterns are negated...

But at least it would be more intuitive from the user perspective.

Reply:
Implemented the new sorting approach and updated docs accordingly. The implementation can be found in the `_sort_value_file_names_for_evaluation` rule helper function